DOCSP-39711 Specify Documents to Return (#57)
mongodb/docs-java-rs: source/specify-documents-to-return.txt

.. _java-rs-specify-documents-to-return:

===========================
Specify Documents to Return
===========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: read, paginate, pagination, order, code example

Overview
--------

In this guide, you can learn how to specify which documents to return from a read
operation by using the following methods:

> - ``limit()``: Specifies the maximum number of documents to return from a query.
> - ``sort()``: Specifies the sort order for the returned documents.
> - ``skip()``: Specifies the number of documents to skip before returning query results.

Reviewer: [issue] Fragments in lists don't need periods per [style guide](https://www.mongodb.com/docs/meta/style-guide/style/lists/#list-items).

```suggestion
- ``limit()``: Specifies the maximum number of documents to return from a query
- ``sort()``: Specifies the sort order for the returned documents
- ``skip()``: Specifies the number of documents to skip before returning query results
```

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``sample_restaurants.restaurants`` collection
from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a

> free MongoDB Atlas cluster and load the sample datasets, see the

Reviewer: [suggestion] Reword for clarity: 
```suggestion
free MongoDB Atlas cluster and load the sample datasets, see
```

:ref:`<java-rs-getting-started>`.

> .. include:: includes/reactor-note.rst

Author: Note: this note has changed. The preview is not accurate: when deployed in, this note will say:

.. important:: Project Reactor Library

   This guide uses the Project Reactor library to consume ``Publisher`` instances returned
   by the {+driver-short+} methods. To learn more about the Project Reactor library
   and how to use it, see `Getting Started <https://projectreactor.io/docs/core/release/reference/#getting-started>`__ in the Reactor documentation. To learn more about how we use Project Reactor
   library methods in this guide, see :ref:`Write Data to MongoDB <write-data-reactor-overview>`.
   
The reason this is not reflected right now is that this is an old branch made before the note changed. It will reflect in the final product.

Limit
-----

To specify the maximum number of documents returned from a read operation,
chain the ``limit()`` method to a ``find()`` method and pass the number
of documents you want to return.

The following example finds all restaurants that have a ``cuisine`` field value

> of ``"Italian"``, and limits the results to 5 documents:

Reviewer: [suggestion] Can style this to use the number 5 as you've done, that way it's consistent with style guide the way you've written out "five" in the next paragraph: 
```suggestion
of ``"Italian"``, and limits the results to ``5`` documents:
```

.. io-code-block::

   .. input:: /includes/specify-documents-to-return.java
      :start-after: start-limit-method
      :end-before: end-limit-method
      :language: java

   .. output::

      Isle Of Capri Resturant Italian
      Arturo'S Italian
      Patsy'S Italian Restaurant Italian
      Piccola Venezia Italian
      Roadhouse Restaurant Italian

.. tip::

   The preceding example returns the first five documents returned by the query, in

>    no particular order. The following section describes how to return the documents

Reviewer: [issue] Can be more specific and link to the glossary description: 
```suggestion
  :manual:`natural order </reference/glossary/#std-term-natural-order>`. The following section describes how to return the documents
```

   in a specified sort order.

.. _java-rs-sort:

Sort
----

> To return documents in a specified order, chain the ``sort()`` method to a

Reviewer: [suggestion] Clarify that there's just one find() method:
```suggestion
To return documents in a specified order, chain the ``sort()`` method to the
```

``find()`` method. The ``sort()`` method takes one parameter: a sort direction

> method. To specify the sort direction, pass either the ``ascending()`` or ``descending``

Reviewer: ```suggestion
method. To specify the sort direction, pass either the ``ascending()`` or ``descending()``
```

> ``find()`` method. The ``sort()`` method takes one parameter: a sort direction
> method. To specify the sort direction, pass either the ``ascending()`` or ``descending``

Reviewer: [suggestion] Potentially reword this to avoid using a [colon](https://www.mongodb.com/docs/meta/style-guide/style/punctuation/#colons) in a sentence: 
```suggestion
``find()`` method. The ``sort()`` method takes the sort direction
method as a parameter. To specify the sort direction, pass either the ``ascending()`` or ``descending``
```

method to the ``sort()`` method. Then, pass the field value you want to sort by.
The ``ascending()`` method sorts values from lowest to highest, and
the ``descending()`` method sorts them from highest to lowest. If you don't specify either

> direction, the method defaults to sorting in ascending order.

Reviewer: [suggestion] Clarify which method just to ensure understanding: 
```suggestion
direction, the ``sort()`` method returns the results in ascending order.
```

The following example returns all documents with the ``cuisine`` value of

> ``"Italian"``, sorted in ascending order:

Reviewer: [suggestion] Refer to the field that's being used: 
```suggestion
``"Italian"``, sorted by the ``name`` field in ascending order:
```

.. io-code-block::

   .. input:: /includes/specify-documents-to-return.java
      :start-after: start-sort-method
      :end-before: end-sort-method
      :language: java

   .. output::

      (Lewis Drug Store) Locanda Vini E Olii
      101 Restaurant And Bar
      44 Sw Ristorante & Bar
      900 Park
      A Voce
      ...
      Zucchero E Pomodori

Skip
----

To skip a specified number of documents before returning your query results,

> chain the ``skip()`` method to a ``find()`` method and pass the number of documents to skip. The

Reviewer: Same as previous suggestion: 
```suggestion
chain the ``skip()`` method to the``find()`` method and pass the number of documents to skip. The
```

``skip()`` method ignores the specified number of documents in your query
results and returns the rest.

The following example returns all documents that have a ``borough`` field value

> of ``"Manhattan"``, and skips the first 10 documents:

Reviewer: If you decide to use monospace for `5` like in the limit example, can also do so here: 
```suggestion
of ``"Manhattan"``, and skips the first ``10`` documents:
```

.. io-code-block::

   .. input:: /includes/specify-documents-to-return.java
      :start-after: start-skip
      :end-before: end-skip
      :language: java

   .. output::

      Cafe Metro
      Lexler Deli

>       Domino'S Pizza

Reviewer: [question] confirming this capitalization is in the data set:
```suggestion
      Domino's Pizza
```

Author: Yup, it's in the data set

return, skipping a specified number of documents before returning.

The following example returns ``5`` documents that have a ``cuisine`` value of
``"Italian"``. The results are sorted in ascending order by ``name`` field value,
skipping the first ``10`` documents.

.. io-code-block::

   .. input:: /includes/specify-documents-to-return.java
      :start-after: start-limit-sort-skip
      :end-before: end-limit-sort-skip
      :language: java

   .. output::
      
      Acqua
      Acqua Restaurant
      Acqua Santa
      Acquista Trattoria
      Acquolina Catering

.. note::

   The order in which you call these methods doesn't change the documents
   that are returned. The driver automatically reorders the calls to perform the
   sort and skip operations first, and the limit operation afterward.

Additional Information
----------------------

.. TODO: uncomment when merged in.
.. For more information about specifying a query, see :ref:`java-rs-specify-query`.

.. For more information about retrieving documents, see :ref:`java-rs-retrieve`.

API Documentation
~~~~~~~~~~~~~~~~~

To learn more about any of the methods or types discussed in this

> guide, see the following API Documentation:

Reviewer: ```suggestion
guide, see the following API documentation:
```

- `find() <{+api+}/mongodb-driver-reactivestreams/com/mongodb/reactivestreams/client/MongoCollection.html#find()>`__

> - `FindPublisher <{+api+}/mongodb-driver-reactivestreams/com/mongodb/reactivestreams/client/FindPublisher.html>`__

Reviewer: [nit] Can potentially move this to the beginning or end of the list so all methods are grouped together.

.. tip::

   The preceding example returns the first five documents returned by the query, in

>    :manual:`natural order </reference/glossary/#std-term-natural-order>`. The
>    following section describes how to return the documents. The following
>    section describes how to return the documents in a specified sort order.

Reviewer: Copy paste typo?
```suggestion
   :manual:`natural order </reference/glossary/#std-term-natural-order>`. The
   following section describes how to return the documents in a specified sort order.
```
